memq-client/src/main/java/com/pinterest/memq/client/commons2/MemqCommonClient.java

```diff
@@ -117,16 +211,21 @@ public CompletableFuture<ResponsePacket> sendRequestPacketAndReturnResponseFutur
       try {
         future = networkClient.send(endpoint.getAddress(), request,
             Duration.ofMillis(timeoutMillis - elapsed));
-        // we keep the endpoint connection for future use
-        currentEndpoint = endpoint;
+        maybeRegisterWriteEndpoint(endpoint, topic);
         break;
       } catch (ExecutionException e) {
         if (e.getCause() instanceof ConnectException) {
           if (retry == retryCount - 1) {
-            logger.error("Failed to send request packet", e);
+            logger.error("Failed to send request packet for topic=" + topic, e);
             throw e;
           } else {
-            logger.warn("Retrying send request after failure: ", e);
+            logger.warn("Retrying send request after failure for topic=" + topic, e);
+            try {
+              deprioritizeDeadEndpoint(endpoint, topic);  // this endpoint is down even after retries in NetworkClient, remove it from the write endpoints and take another one from locality endpoints
+            } catch (Exception ex) {
+              logger.error("Failed to refresh write endpoints", ex);
+              throw e;
```

Comment:
Should it throw `ex` too?

Reply:
we're throwing e instead of ex because ex is a secondary error thrown by `deprioritizeDeadEndpoint()`. the main issue is the `ConnectException`.